google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingService.java

```diff
@@ -0,0 +1,284 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.logging;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.gax.core.ApiFuture;
+import com.google.api.gax.core.ApiFutureCallback;
+import com.google.api.gax.core.ApiFutures;
+import com.google.cloud.MonitoredResource;
+import com.google.cloud.logging.Logging.WriteOption;
+import com.google.common.util.concurrent.Uninterruptibles;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Set;
+
+public class LoggingService {
+
+  private static final ThreadLocal<Boolean> inPublishCall = new ThreadLocal<>();
+  private volatile Logging logging;
+  private LoggingOptions loggingOptions;
+  private WriteOption[] writeOptions;
+  private Severity flushSeverity;
+  private Synchronicity synchronicity;
+  private LoggingErrorHandler errorHandler;
+  private List<LoggingEnhancer> resourceEnhancers;
+
+  private final Object writeLock = new Object();
+  private final Set<ApiFuture<Void>> pendingWrites =
+      Collections.newSetFromMap(new IdentityHashMap<ApiFuture<Void>, Boolean>());
+
+  private LoggingService() {}
+
+  public void flush() {
+    // BUG(1795): We should force batcher to issue RPC call for buffered messages,
+    // so the code below doesn't wait uselessly.
+    ArrayList<ApiFuture<Void>> writesToFlush = new ArrayList<>();
+    synchronized (writeLock) {
+      writesToFlush.addAll(pendingWrites);
+    }
+    for (ApiFuture<Void> write : writesToFlush) {
+      try {
+        Uninterruptibles.getUninterruptibly(write);
+      } catch (Exception e) {
+        // Ignore exceptions, they are propagated to the error manager.
+      }
+    }
+  }
+
+  private void enhanceLogEntry(LogEntry.Builder entryBuilder) {
+    for (LoggingEnhancer enhancer : resourceEnhancers) {
+      enhancer.enhanceLogEntry(entryBuilder);
+    }
+  }
+
+  public void publish(LogEntry.Builder entryBuilder) {
+    enhanceLogEntry(entryBuilder);
+    LogEntry entry = entryBuilder.build();
+
+    if (inPublishCall.get() != null) {
+      // ignore all logs generated in the course of logging through this handler
+      return;
+    }
+    inPublishCall.set(true);
+
+    try {
+      write(entry);
+      if (entry.getSeverity().compareTo(flushSeverity) >= 0) {
+        flush();
+      }
+    } finally {
+      inPublishCall.remove();
+    }
+  }
+
+  /** Closes the handler and the associated {@link Logging} object. */
+  public synchronized void close() throws SecurityException {
+    if (logging != null) {
+      try {
+        logging.close();
+      } catch (Exception ex) {
+        // ignore
+      }
+    }
+    logging = null;
+  }
+
+  public synchronized void setFlushSeverity(Severity severity) {
+    flushSeverity = severity;
+  }
```

Comment:
Agree, i missed seeing the mutability here. I've updated to new and removed the builder pattern.